### PR TITLE
Assignment4 - Josh Pedersen & Conner Hanson

### PR DESCRIPTION
**Pull request by Joshua-Pedersen**
Partner: @connerhanson

### CODE REVIEW

File: Logger/Person.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Logger;
+
+public record class Person(FullName fullName) : BaseEntity
+{
+    // Implicit as we are passing in the fullname to be used.
+    public override string Name { get => fullName.Name; }
+    
+}
```
Comment (omar5o9) [inline Logger/Person.cs:8]: Smart idea for making a person class to share the code in student and employee

**Review by omar5o9 — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by Anna-Mclenko** [comment]
Also Reviewing

File: Logger.Tests/EntitiesTests.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        // Arrange
+        FullName testName = new("John", "Doe");
+
+        // Act
+        Student student1 = new(testName);
+        Student student2 = new(testName) { Id = student1.Id};
+
+        // Assert
+        Assert.Equal(student1, student2);
```
Comment (Anna-Mclenko) [inline Logger.Tests/EntitiesTests.cs:63]: I thought we were supposed to test the equal method that the record implements behind the scenes. 
like `Assert.True(student1.Equals(student2));`

but I guess thats another way to test it :)

**Review by Anna-Mclenko — COMMENTED** [review]
Tasks

Define a full name record (first, last, middle) handling optional and null appropriately. ✔

Provide a comment on the full name record on why you selected to define a value or a
reference type and ✔

Provide a comment on the full name record on why or why not the type is immutable. ✔

Define an IEntity interface: ✔

Add an Id property of type Guid that is an init-only setter. ✔

Add a Name property that is string. ✔

Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔

Do not implement the Name property in this abstract class. ✔

Do force any derived classes to provide an implementation for Name. ✔

Define book, student, and employee records - all with entity capabilities. ✔

Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔

Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔

You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔

Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔

Test that the equality behavior on these entities behaves as expected. ✔

Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.✔

Unit test the Storage class using various entities to ensure that there are no bugs.✔

Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.

Fundamentals

Be sure you enable:
Enable Nullability for all projects ✔

Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔

Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔

Ensure that you turn on EnforceCodeStyleInBuild ✔

Ensure that you turn on Nullability (Nullable) ✔

Ensure there are no errors or warnings (including code analysis warnings) ✔

All of the above should be unit tested ✔

Choose simplicity over complexity ✔

Extra Credit

Refactor your project files to put common properties and/or items into a single file. ❌

Convert ILogger.CreateLogger to a generic method. ❌

Returns the logger classes type that is to be created. ❌

Has a type parameter for the configuration type ❌

File: Logger/BaseEntity.cs
Comment (HadiM2) [inline Logger/BaseEntity.cs:1]: Good job on making it abstract, this way its forced on any derived class.

File: Logger/Entities.cs
Comment (HadiM2) [inline Logger/Entities.cs:1]: Great job! this never occurred to me, you guys are the first people I've seen combine them into one place, smart.

File: Logger/FullName.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    
+}
+
+// I decided to define FullName as a value type because it repersents a single entity with few attributes.
+// From what I know value types seem to be the better option in terms of memory usage for minimal objects. 
+
+// As far as immutability goes I chose to make them immutable since in a real world scenario having 
+// muttable names may create a vulnerability within the code. I also know from another class i've taken
+// that muttable objects are not thread safe, though I don't believe that would be a problem for this
```
Comment (HadiM2) [inline Logger/FullName.cs:27]: I also remember this being mentioned. great reasonings all around.

File: Logger/IEntity.cs [hunk 4]
```diff
@@ -1,7 +1,6 @@
 ﻿namespace Logger;
 public interface IEntity
 {
-
-        // Place members here.
-
+	Guid Id { get; init; }
+	string Name { get; }
```
Comment (HadiM2) [inline Logger/IEntity.cs:5]: You didn't have a need for a setter?

File: Logger/Logger.csproj
Comment (HadiM2) [inline Logger/Logger.csproj:1]: ✔

File: Logger/Person.cs
Comment (HadiM2) [inline Logger/Person.cs:1]: I like how simple you guys made it, good shit. ✔

File: Logger.Tests/StorageTests.cs
Comment (HadiM2) [inline Logger.Tests/StorageTests.cs:1]: Remove_Student_DoesNotContainStudent() is a great test idea, I didn't think about testing that feature.

File: Logger.Tests/Logger.Tests.csproj
Comment (HadiM2) [inline Logger.Tests/Logger.Tests.csproj:1]: ✔

File: Logger.Tests/FullNameTests.cs
Comment (HadiM2) [inline Logger.Tests/FullNameTests.cs:1]: ✔

File: Logger.Tests/EntitiesTests.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        // Act
+        Book book1 = new(testTitle);
+        Book book2 = new(testTitle) { Id = book1.Id };
+
+        // Assert
+        Assert.Equal(book1, book2);
+    }
+
+    [Fact]
+    public void Book_SameNameDiffId_NotEqual()
```
Comment (HadiM2) [inline Logger.Tests/EntitiesTests.cs:123]: Good job testing on reference which should be not equal even if the data is actually equal. ✔

**Review by HadiM2 — COMMENTED** [review]
Tasks
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔ **- Employee and student are done through person class**
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔ **- Person class**
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔ **- Book doesn't use it**
Test that the equality behavior on these entities behaves as expected. ✔**Annas suggestion is good : )**
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly. ✔
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.) ✔

Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔ **- Combining record classes.**

File: Logger/BaseEntity.cs
Comment (connerhanson) [reply Logger/BaseEntity.cs:1]: Our thoughts exactly, thanks!

File: Logger/Entities.cs
Comment (connerhanson) [reply Logger/Entities.cs:1]: Just trying to keep it minimal.

File: Logger/IEntity.cs (hunk 4 above)
Comment (connerhanson) [reply Logger/IEntity.cs:5]: We push the setter implementation down to the entitles themself, so for example the student/employee/person can use FullName while the book can just use a standard string name.

File: Logger.Tests/StorageTests.cs
Comment (connerhanson) [reply Logger.Tests/StorageTests.cs:1]: Thanks!

File: Logger.Tests/EntitiesTests.cs (hunk 5 above)
Comment (connerhanson) [reply Logger.Tests/EntitiesTests.cs:123]: Thanks!

File: Logger.Tests/EntitiesTests.cs (hunk 2 above)
Comment (Joshua-Pedersen) [reply Logger.Tests/EntitiesTests.cs:63]: Most tests the better, thanks for the recommendation

File: Logger.Tests/EntitiesTests.cs [hunk 6]
```diff
@@ -0,0 +1,219 @@
+﻿using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Xunit;
+
+namespace Logger.Tests;
+
+public class EntitiesTests
+{
+
+    [Fact]
+    public void Student_GivenName_NameMatches()
```
Comment (BenjaminMichaelis) [inline Logger.Tests/EntitiesTests.cs:10]: All student tests should be in StudentTests.cs and employee tests be in EmployeeTests.cs and etc.

File: Logger.Tests/EntitiesTests.cs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        // Act
+        Student student1 = new(testName);
+        Student student2 = new(testName) { Id = student1.Id};
+
+        // Assert
+        Assert.Equal(student1, student2);
+    }
+
+    [Fact]
+    public void Student_MatchingProperties_TrueEquality()
```
Comment (BenjaminMichaelis) [inline Logger.Tests/EntitiesTests.cs:67]: What value is this test adding? And the following ones?

File: Logger.Tests/FullNameTests.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	[Fact]
+	public void FullName_CreationWithNullLastName_ThrowsException()
+	{
+		// Arrange
+		string firstName = "firstName";
+
+		// Act & Assert
+		Assert.Throws<ArgumentNullException>(() => new FullName(firstName, null!));
```
Comment (BenjaminMichaelis) [inline Logger.Tests/FullNameTests.cs:60]: A lot of the variables in these tests could be inlined
```suggestion
		// Act & Assert
		Assert.Throws<ArgumentNullException>(() => new FullName("firstName", null!));
```

File: Logger.Tests/StorageTests.cs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    {
+        // Arrange
+        Storage storage = new();
+        Student student = new(new FullName("First", "Last"));
+
+        // Act
+        storage.Add(student);
+
+        // Assert
+        Assert.True(storage.Contains(student));
```
Comment (BenjaminMichaelis) [inline Logger.Tests/StorageTests.cs:18]: Why not use `Assert.Contains()` here, it seems better than Assert.True a contains

File: Logger.Tests/StorageTests.cs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        // Arrange
+        Storage storage = new();
+        Student student = new(new FullName("First", "Last"));
+
+        // Act
+        storage.Add(student);
+        storage.Remove(student);
+
+        // Assert
+        Assert.False(storage.Contains(student));
```
Comment (BenjaminMichaelis) [inline Logger.Tests/StorageTests.cs:61]: `Assert.DoesNotContain()` seems better here

File: Logger/BaseEntity.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+namespace Logger;
+
+public abstract record class BaseEntity : IEntity
+{
+    public Guid Id { get; init; } = Guid.NewGuid();
```
Comment (BenjaminMichaelis) [inline Logger/BaseEntity.cs:5]: Is there ever a time the Id is set by the calling class? 
```suggestion
    public Guid Id { get; } = Guid.NewGuid();
```

File: Logger/Entities.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Logger;
+
+public record class Book(string title) : BaseEntity
```
Comment (BenjaminMichaelis) [inline Logger/Entities.cs:3]: Classes should each be in their own file

Comment (BenjaminMichaelis) [inline Logger/Entities.cs:3]: I think there are other valuable properties to a Book, like the Author, and Title, etc.

File: Logger/FullName.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
+using System;
+
+namespace Logger;
+
+public readonly record struct FullName
```
Comment (BenjaminMichaelis) [inline Logger/FullName.cs:5]: struct is a good decision here

File: Logger/Person.cs [hunk 14]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Logger;
+
+public record class Person(FullName fullName) : BaseEntity
+{
+    // Implicit as we are passing in the fullname to be used.
+    public override string Name { get => fullName.Name; }
```
Comment (BenjaminMichaelis) [inline Logger/Person.cs:6]: This should have at a minimum a `FirstName` and `LastName`

File: Logger/Entities.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+{
+    // Implicit due to the fact we are passing the name/title in
+    public override string Name { get; } = title;
+}
+public record class Student : Person
+{
+    // Explicit as Student's name comes from the Person class
+    public Student(FullName name) : base(name) { }
+}
+public record class Employee : Person
```
Comment (BenjaminMichaelis) [inline Logger/Entities.cs:13]: Employee's (and students) I can see having their own Id's. I encourage you to not just do the bare minimum to meet requirements.

**Pull request closed without merging**